### ISSUE TITLE
Automatic verification of untrusted contributors adding crates which are trusted by imported sources

### ISSUE THREAD

**Issue by alexcrichton** [mozilla/cargo-vet#271]
Some Wasmtime folks had a great discussion with @bholley yesterday about the usage of `cargo vet` in Wasmtime and the Bytecode Alliance in general. One of the discussion points that came up was a scenario in Wasmtime which we ideally would like to have handled automatically but sounds like would need verification today. Our scenario is:

* An "untrusted" contributor sends a PR (not necessarily malicious, just one we aren't previously aware of) which adds a new dependency.
* The `cargo vet` run on CI fails so the contributor runs `cargo vet` locally. Assuming Wasmtime imports audits from Firefox there could be a situation where Firefox has provided an audit of this crate in which case `imports.lock` is updated.
* The contributor commits the changes to `imports.lock` and CI starts passing.

In this situation as a reviewer of the PR we don't actually know if the changes to `imports.lock` are "trusted". In this situation it sounded like we would be required to checkout the PR locally and verify that our own run of `cargo vet` produced no changes to `imports.lock`. Ideally though we would like CI to report this failure if it could.

We had some various discussion about possibilities here but I don't think that any firm conclusion was reached about how possibly to solve this. In any case though I wanted to open up an issue to have a good place to discuss!

**Comment by mystor** [comment]
Yeah, right now if you're running with `--locked` like you likely would in CI, the server would not make any attempt to verify that the state of imports.lock remains in-sync with the remote audits which you're fetching, and it's a bit difficult to do that in a fully-reliable way, because we're currently using http to fetch remote audits, meaning that there's no history, and the page could update at any time, so we can't actually check that at some point (presumably when the contributor ran `cargo vet`) the remote audits matched the given ones.

I believe there were some discussions at one point about using something like a git repository in order to track the history of imported audits. If we do something like that, we could include some information about what commit the import came from upstream which you could use to verify the change either manually or with some automated system.

Another option here could be using checksums. Once #261 lands, we could add a SHA256 hash of the imported audits file into the fetched metadata which would be validated when we read imports.lock, and could be checked against some log of valid SHA256 hashes hosted somewhere by your upstream source, which could contain the last N days of published audits.toml hashes (though this may require some additional tooling to make workable in a broad sense).

If we're OK with spuriously failing some check if the upstream audits changed between the contributor running `cargo vet` and the server running it, we could also potentially include a flag like `--verify-imports` which errors out if the imports would change at all, but I worry that some upstream sources may change too rapidly for that to be a reasonable thing to require. Perhaps it could fetch the upstream source and also run `cargo vet` against the updated sources, making sure that whatever audits you're using would still pass with whatever the most up-to-date sources are?

**Comment by bholley** [comment]
I think we can achieve the specific goal of verifying untrusted imports.lock changes as follows:
(1) Add a `cargo vet verify-cached-imports` which refetches all the imports and succeeds if every audit in `imports.lock` also exists upstream. It would not fail if new audits have been added upstream.
(2) Add an additional CI job to run `verify-cached-imports` that is only triggered if imports.lock has changed. This would ensure that it runs when `imports.lock` is actually modified, but prevent it from running (and potentially failing) on arbitrary unrelated PRs.

Now, there is an interesting design choice as to whether step (2) is actually desirable (is determinism more valuable than faster detection of revoked audits, which we expect to be rare). But we can discuss that over in #272.

**Comment by jameysharp** [comment]
I don't think we'd want to detect revoked audits during unrelated pull requests, but I think it makes sense to run a periodic CI task for that. Whether an audit was removed from some imported collection of audits, or a violation has been added, either way I'd want to detect that in a timely fashion and independently of anybody opening new pull requests.

So I think (2) is unambiguously a good idea. If somebody has submitted a changeset for review where they needed to update `imports.lock` to get `cargo vet` to pass, then there should be a short time period between when they ran `cargo vet` and when CI checks it. If `verify-cached-imports` only runs on changes to `imports.lock`, that limits the window where they're racing with upstream audit changes. And in the unlikely event that, within that narrow window, upstream invalidates the results the contributor saw locally from `cargo vet`, at least their run of `cargo vet` should be fresh in their memory and they'll know how to fix it.

I think that's probably sufficient. But here are a couple other ways that `verify-cached-imports` could make failures even less likely for non-malicious contributions:

- Examine a diff, rather than the entire `imports.lock`, and only check that the newly-added entries also exist upstream. Entries removed locally either don't matter because they're not used, or the separate `cargo vet --locked` step will catch them. And if an audit is invalidated upstream, but we had already imported it, that isn't the contributor's fault and shouldn't reflect poorly on their PR; it should be caught by a separate periodic CI check.

- Only check that audits which are needed to validate the current `Cargo.lock` are also present upstream. If an audit gets invalidated but we don't need it, why worry?

Just don't use _both_ of those options at once, or somebody can submit a fake audit in `imports.lock` at one point without being detected since it isn't used, and then later change `Cargo.lock` to depend on that audit without being detected since it isn't new.

**Comment by bholley** [comment]
If we proceed with the proposal in #272 to only cache audits that are actually required, then `verify-cached-imports` will only fail if either (A) a contributor lied, or (B) an imported audit _which this project was specifically depending on_ was revoked.

Both of these are exceedingly rare, but things we want to catch quickly. Because it's rare, it would probably work fine in practice to just bundle this checking into `cargo vet --locked`, but that would violate the principle of determinism, which is a useful principle to respect.

Therefore, I think the optimal solution here is to run `verify-cached-imports` whenever `imports.lock` is changed (which would catch (A)), as well as on a cron job (which would catch (B)). We can provide some GH action templates for people to crib to do this.

**Comment by bholley** [comment]
> Hmm, what would y'all think about solving this problem through dependabot?

Per Matrix discussion, I'd rather avoid entangling dependabot here, and see a number of ancillary benefits to minimizing imports.lock beyond just the narrow consideration of detecting fraud. If we're willing to err on the side of including "possibly-relevant" audits, we should be able to bypass most of the tricky engineering challenges with identifying a truly-minimal set.

**Comment by mystor** [comment]
This should probably be easier to do with the new handling of imports, so long as your upstreams do not revoke audits by removing them from the audit file (as that could cause this check to fail, but that may be desirable). We could have a helper subcommand which would fetch live audit sources, and check that every audit which we have recorded from that source is also present in the remote source, erroring out if any of the recorded audits are missing (e.g. because they were falsified or were revoked). We could also potentially report new violation conflicts and criteria description changes in that situation which may have changed since the live version, though we might want that behaviour to be optional.

Hopefully people aren't revoking audits frequently enough for that potential race to be an issue here, and getting a CI failure in that case may actually be desirable.

@alexcrichton would a subcommand like that be sufficient for wasmtime to help counter this issue? Hopefully #296 will have reduced the churn substantially, so it may not be necessary anymore, but it may also be a useful tool to run in CI anyway.

**Comment by alexcrichton** [comment]
It might be best to see how #296 integrates into Wasmtime and what our experiences are with it. I suspect we may still need this but it's probably best to get concrete use cases as to why we feel we need it to help guide design of a possible new subcommand. Thanks for the work here on this though!
